// @botonic/plugin-contentful/lib/contentful/contents/carousel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CarouselDelivery = void 0;
const cms = __importStar(require("../../cms"));
const delivery_utils_1 = require("../delivery-utils");
const follow_up_1 = require("./follow-up");
// TODO does not yet load the followU p
class CarouselDelivery extends follow_up_1.DeliveryWithFollowUp {
    constructor(delivery, button, resumeErrors) {
        super(cms.ContentType.CAROUSEL, delivery, resumeErrors);
        this.button = button;
    }
    carousel(id, context) {
        return __awaiter(this, void 0, void 0, function* () {
            const entry = yield this.getEntry(id, context);
            return this.fromEntry(entry, context);
        });
    }
    fromEntry(entry, context) {
        return __awaiter(this, void 0, void 0, function* () {
            const elements = yield this.asyncMap(context, entry.fields.elements, (entry) => __awaiter(this, void 0, void 0, function* () {
                return this.elementFromEntry(entry, context);
            }));
            return delivery_utils_1.addCustomFields(new cms.Carousel(delivery_utils_1.ContentfulEntryUtils.commonFieldsFromEntry(entry), elements), entry.fields);
        });
    }
    element(id, context) {
        return __awaiter(this, void 0, void 0, function* () {
            const entry = yield this.delivery.getEntry(id, context);
            return this.elementFromEntry(entry, context);
        });
    }
    elementFromEntry(entry, context) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            this.checkEntry(entry);
            const fields = entry.fields;
            const buttonEntries = fields.buttons || [];
            const buttons = yield this.button.fromReferenceSkipErrors(buttonEntries, context);
            return new cms.Element(entry.sys.id, buttons, (_a = fields.title) !== null && _a !== void 0 ? _a : '', (_b = fields.subtitle) !== null && _b !== void 0 ? _b : '', this.urlFromAssetOptional(fields.pic, context));
        });
    }
}
exports.CarouselDelivery = CarouselDelivery;
