// @botonic/plugin-contentful/lib/markup/whatsapp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhatsApp = void 0;
const markup_1 = require("./markup");
class WhatsApp extends markup_1.MarkUp {
    constructor() {
        super(markup_1.MarkupType.WHATSAPP);
        this.type = markup_1.MarkupType.WHATSAPP;
        this.STRONG = '*';
        this.EMPHASIS = '_';
    }
    parse(txt) {
        throw new Error('WhatsApp.parse not implemented');
    }
    renderToken(token) {
        if (token.items) {
            return token.items.map(item => this.render(item.tokens)).join('\n');
        }
        const inner = token.tokens
            ? this.render(token.tokens, '')
            : token.text || '';
        if (token.type == markup_1.TokenType.STRONG) {
            return `${this.STRONG}${inner}${this.STRONG}`;
        }
        if (token.type == markup_1.TokenType.EMPHASIS) {
            return `${this.EMPHASIS}${inner}${this.EMPHASIS}`;
        }
        return inner || '';
    }
    wrapWithInline(input, inlineType) {
        if (inlineType === markup_1.TokenType.EMPHASIS) {
            return `_${input}_`;
        }
        if (inlineType === markup_1.TokenType.STRONG) {
            return `*${input}*`;
        }
        throw new Error(`wrapWithInline does not support inline ${inlineType}`);
    }
}
exports.WhatsApp = WhatsApp;
